File: dracpond/src/main/java/cc/chordflower/dracpond/utils/ProxyFile.java
package cc.chordflower.dracpond.utils;

import org.jetbrains.annotations.NotNull;

import java.io.File;
import java.io.FileFilter;
import java.io.FilenameFilter;
import java.io.IOException;
import java.net.URI;
import java.nio.file.FileAlreadyExistsException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.attribute.FileTime;
import java.nio.file.attribute.PosixFilePermission;
import java.util.EnumSet;
import java.util.HashSet;
import java.util.Set;

/**
 * Implements the file api using the given path instead, to allow for diferent filesystem implementations.
 */
@SuppressWarnings( { "unchecked", "ResultOfMethodCallIgnored", "unused" } )
public class ProxyFile extends File {

  private static final long serialVersionUID = 1L;
  private final Path _filePath;

  public ProxyFile( Path filePath ) {
    super( "/" );
    this._filePath = filePath.normalize( );
  }

  @Override
  public boolean canExecute( ) {
    return Files.isExecutable( this._filePath );
  }

  @Override
  public boolean canRead( ) {
    return Files.isReadable( this._filePath );
  }

  @Override
  public boolean canWrite( ) {
    return Files.isWritable( this._filePath );
  }

  @Override
  public int compareTo( File pathname ) {
    return this._filePath.compareTo( pathname.toPath( ) );
  }

  @Override
  public boolean createNewFile( ) throws IOException {
    try {
      Files.createFile( this._filePath );
      return true;
    } catch( FileAlreadyExistsException ex ) {
      return false;
    }
  }

  @Override
  public boolean delete( ) {
    try {
      Files.delete( this._filePath );
      return true;
    } catch( IOException ex ) {
      return false;
    }
  }

  @Override
  public void deleteOnExit( ) {
    this.delete( );
  }

  @Override
  public boolean equals( Object obj ) {
    if( obj instanceof File ) {
      return this._filePath.equals( ( ( File )obj ).toPath( ) );
    }
    return false;
  }

  @Override
  public boolean exists( ) {
    return Files.exists( this._filePath );
  }

  @Override
  public @NotNull File getAbsoluteFile( ) {
    return new ProxyFile( this._filePath.toAbsolutePath( ) );
  }

  @Override
  public @NotNull String getAbsolutePath( ) {
    return this._filePath.toAbsolutePath( ).toString( );
  }

  @Override
  public @NotNull File getCanonicalFile( ) {
    return new ProxyFile( this._filePath.toAbsolutePath( ).normalize( ) );
  }

  @Override
  public @NotNull String getCanonicalPath( ) {
    return this._filePath.toAbsolutePath( ).normalize( ).toString( );
  }

  @Override
  public long getFreeSpace( ) {
    return 0L;
  }

  @Override
  public @NotNull String getName( ) {
    return this._filePath.getFileName( ).toString( );
  }

  @Override
  public String getParent( ) {
    return this._filePath.toAbsolutePath( ).normalize( ).getParent( ).toString( );
  }

  @Override
  public File getParentFile( ) {
    return new ProxyFile( this._filePath.toAbsolutePath( ).normalize( ).getParent( ) );
  }

  @Override
  public @NotNull String getPath( ) {
    return this._filePath.toString( );
  }

  @Override
  public long getTotalSpace( ) {
    return 0L;
  }

  @Override
  public long getUsableSpace( ) {
    return 0L;
  }

  @Override
  public int hashCode( ) {
    return this._filePath.hashCode( );
  }

  @Override
  public boolean isAbsolute( ) {
    return this._filePath.isAbsolute( );
  }

  @Override
  public boolean isDirectory( ) {
    return Files.isDirectory( this._filePath );
  }

  @Override
  public boolean isFile( ) {
    return Files.isRegularFile( this._filePath );
  }

  @Override
  public boolean isHidden( ) {
    try {
      return Files.isHidden( this._filePath );
    } catch( IOException e ) {
      return false;
    }
  }

  @Override
  public long lastModified( ) {
    try {
      return Files.getLastModifiedTime( this._filePath ).toMillis( );
    } catch( IOException e ) {
      return 0L;
    }
  }

  @Override
  public long length( ) {
    try {
      return Files.size( this._filePath );
    } catch( IOException e ) {
      return 0L;
    }
  }

  @Override
  public String[ ] list( ) {
    try {
      return Files.list( this._filePath ).map( Object::toString ).toArray( String[ ]::new );
    } catch( IOException e ) {
      return null;
    }
  }

  @Override
  public String[ ] list( FilenameFilter filter ) {
    try {
      return Files.list( this._filePath ).filter( path -> filter.accept( this, path.getFileName( ).toString( ) ) ).map( Object::toString ).toArray( String[ ]::new );
    } catch( IOException e ) {
      return null;
    }
  }

  @Override
  public File[ ] listFiles( ) {
    try {
      return Files.list( this._filePath ).map( ProxyFile::new ).toArray( File[ ]::new );
    } catch( IOException e ) {
      return null;
    }
  }

  @Override
  public File[ ] listFiles( FileFilter filter ) {
    try {
      if( filter == null ) {
        throw new NullPointerException( );
      }
      return Files.list( this._filePath ).map( ProxyFile::new ).filter( filter::accept ).toArray( File[ ]::new );
    } catch( IOException e ) {
      return null;
    }
  }

  @Override
  public File[ ] listFiles( FilenameFilter filter ) {
    try {
      return Files.list( this._filePath ).filter( path -> filter.accept( this, path.getFileName( ).toString( ) ) ).map( ProxyFile::new ).toArray( File[ ]::new );
    } catch( IOException e ) {
      return null;
    }
  }

  @Override
  public boolean mkdir( ) {
    try {
      Files.createDirectory( this._filePath );
      return true;
    } catch( IOException e ) {
      return false;
    }
  }

  @Override
  public boolean mkdirs( ) {
    try {
      return Files.createDirectories( this._filePath ) != null;
    } catch( IOException e ) {
      return false;
    }
  }

  @Override
  public boolean renameTo( File dest ) {
    try {
      Files.move( this._filePath, dest.toPath( ) );
      return true;
    } catch( IOException e ) {
      return false;
    }
  }

  @Override
  public boolean setExecutable( boolean executable ) {
    return this.setExecutable( executable, true );
  }

  @Override
  public boolean setExecutable( boolean executable, boolean ownerOnly ) {
    try {
      Set< PosixFilePermission > permissions = new HashSet<>( ( Set< PosixFilePermission > )Files.getAttribute( this._filePath, "posix.permissions" ) );
      if( executable ) {
        EnumSet< PosixFilePermission > newPermissions = ownerOnly ? EnumSet.of( PosixFilePermission.OWNER_EXECUTE ) : EnumSet.of( PosixFilePermission.GROUP_EXECUTE, PosixFilePermission.OTHERS_EXECUTE, PosixFilePermission.OWNER_EXECUTE );
        permissions.addAll( newPermissions );
      } else {
        permissions.remove( PosixFilePermission.OWNER_EXECUTE );
        if( !ownerOnly ) {
          permissions.remove( PosixFilePermission.GROUP_EXECUTE );
          permissions.remove( PosixFilePermission.OTHERS_EXECUTE );
        }
      }
      Files.setAttribute( this._filePath, "posix.permissions", permissions );
      return true;
    } catch( IOException e ) {
      return false;
    }
  }

  @Override
  public boolean setLastModified( long time ) {
    try {
      Files.setLastModifiedTime( this._filePath, FileTime.fromMillis( time ) );
      return true;
    } catch( IOException e ) {
      return false;
    }
  }

  @Override
  public boolean setReadable( boolean readable ) {
    return this.setReadable( readable, true );
  }

  @Override
  public boolean setReadable( boolean readable, boolean ownerOnly ) {
    try {
      Set< PosixFilePermission > permissions = new HashSet<>( ( Set< PosixFilePermission > )Files.getAttribute( this._filePath, "posix.permissions" ) );
      if( readable ) {
        EnumSet< PosixFilePermission > newPermissions = ownerOnly ? EnumSet.of( PosixFilePermission.OWNER_READ ) : EnumSet.of( PosixFilePermission.GROUP_READ, PosixFilePermission.OTHERS_READ, PosixFilePermission.OWNER_READ );
        permissions.addAll( newPermissions );
      } else {
        permissions.remove( PosixFilePermission.OWNER_READ );
        if( !ownerOnly ) {
          permissions.remove( PosixFilePermission.GROUP_READ );
          permissions.remove( PosixFilePermission.OTHERS_READ );
        }
      }
      Files.setAttribute( this._filePath, "posix.permissions", permissions );
      return true;
    } catch( IOException e ) {
      try {
        Files.setAttribute( this._filePath, "dos.readonly", readable );
        return true;
      } catch( IOException e1 ) {
        return false;
      }
    }
  }

  @Override
  public boolean setReadOnly( ) {
    try {
      Set< PosixFilePermission > newPermissions = EnumSet.of( PosixFilePermission.GROUP_READ, PosixFilePermission.OTHERS_READ, PosixFilePermission.OWNER_READ );
      Set< PosixFilePermission > permissions = new HashSet<>( ( Set< PosixFilePermission > )Files.getAttribute( this._filePath, "posix.permissions" ) );
      permissions.addAll( newPermissions );
      Files.setAttribute( this._filePath, "posix.permissions", permissions );
      return true;
    } catch( IOException e ) {
      try {
        Files.setAttribute( this._filePath, "dos.readonly", Boolean.TRUE );
        return true;
      } catch( IOException e1 ) {
        return false;
      }
    }
  }

  @Override
  public boolean setWritable( boolean writable ) {
    return this.setWritable( writable, true );
  }

  @Override
  public boolean setWritable( boolean writable, boolean ownerOnly ) {
    try {
      Set< PosixFilePermission > permissions = new HashSet<>( ( Set< PosixFilePermission > )Files.getAttribute( this._filePath, "posix.permissions" ) );
      if( writable ) {
        EnumSet< PosixFilePermission > newPermissions = ownerOnly ? EnumSet.of( PosixFilePermission.OWNER_WRITE ) : EnumSet.of( PosixFilePermission.GROUP_WRITE, PosixFilePermission.OTHERS_WRITE, PosixFilePermission.OWNER_WRITE );
        permissions.addAll( newPermissions );
      } else {
        permissions.remove( PosixFilePermission.OWNER_WRITE );
        if( !ownerOnly ) {
          permissions.remove( PosixFilePermission.GROUP_WRITE );
          permissions.remove( PosixFilePermission.OTHERS_WRITE );
        }
      }
      Files.setAttribute( this._filePath, "posix.permissions", permissions );
      return true;
    } catch( IOException e ) {
      return false;
    }
  }

  @Override
  public @NotNull Path toPath( ) {
    return this._filePath;
  }

  @Override
  public String toString( ) {
    return this._filePath.toString( );
  }

  @Override
  public @NotNull URI toURI( ) {
    return this._filePath.toUri( );
  }

}

File: dracpond/src/main/kotlin/cc/chordflower/dracpond/utils/ProxyFile.kt
@file:Suppress("UNCHECKED_CAST")

package cc.chordflower.dracpond.utils

import java.io.File
import java.io.FileFilter
import java.io.FilenameFilter
import java.io.IOException
import java.net.URI
import java.nio.file.FileAlreadyExistsException
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.attribute.FileTime
import java.nio.file.attribute.PosixFilePermission
import java.util.*

/**
 * Implements the file api using the given path instead, to allow for diferent filesystem implementations.
 */
class ProxyFile(filePath: Path) : File("/") {
  private val _filePath: Path = filePath.normalize()
  override fun canExecute(): Boolean {
    return Files.isExecutable(_filePath)
  }

  override fun canRead(): Boolean {
    return Files.isReadable(_filePath)
  }

  override fun canWrite(): Boolean {
    return Files.isWritable(_filePath)
  }

  override fun compareTo(other: File): Int {
    return _filePath.compareTo(other.toPath())
  }

  @Throws(IOException::class) override fun createNewFile(): Boolean {
    return try {
      Files.createFile(_filePath)
      true
    } catch(ex: FileAlreadyExistsException) {
      false
    }
  }

  override fun delete(): Boolean {
    return try {
      Files.delete(_filePath)
      true
    } catch(ex: IOException) {
      false
    }
  }

  override fun deleteOnExit() {
    delete()
  }

  override fun equals(other: Any?): Boolean {
    return if(other is File) {
      _filePath == other.toPath()
    } else false
  }

  override fun exists(): Boolean {
    return Files.exists(_filePath)
  }

  override fun getAbsoluteFile(): File {
    return ProxyFile(_filePath.toAbsolutePath())
  }

  override fun getAbsolutePath(): String {
    return _filePath.toAbsolutePath().toString()
  }

  override fun getCanonicalFile(): File {
    return ProxyFile(_filePath.toAbsolutePath().normalize())
  }

  override fun getCanonicalPath(): String {
    return _filePath.toAbsolutePath().normalize().toString()
  }

  override fun getFreeSpace(): Long {
    return 0L
  }

  override fun getName(): String {
    return _filePath.fileName.toString()
  }

  override fun getParent(): String {
    return _filePath.toAbsolutePath().normalize().parent.toString()
  }

  override fun getParentFile(): File {
    return ProxyFile(_filePath.toAbsolutePath().normalize().parent)
  }

  override fun getPath(): String {
    return _filePath.toString()
  }

  override fun getTotalSpace(): Long {
    return 0L
  }

  override fun getUsableSpace(): Long {
    return 0L
  }

  override fun hashCode(): Int {
    return _filePath.hashCode()
  }

  override fun isAbsolute(): Boolean {
    return _filePath.isAbsolute
  }

  override fun isDirectory(): Boolean {
    return Files.isDirectory(_filePath)
  }

  override fun isFile(): Boolean {
    return Files.isRegularFile(_filePath)
  }

  override fun isHidden(): Boolean {
    return try {
      Files.isHidden(_filePath)
    } catch(e: IOException) {
      false
    }
  }

  override fun lastModified(): Long {
    return try {
      Files.getLastModifiedTime(_filePath).toMillis()
    } catch(e: IOException) {
      0L
    }
  }

  override fun length(): Long {
    return try {
      Files.size(_filePath)
    } catch(e: IOException) {
      0L
    }
  }

  override fun list(): Array<String> {
    return try {
      Files.list(_filePath).map( { obj: Path -> obj.toString() } ).toArray( { arrayOf() } )
    } catch(e: IOException) {
      arrayOf()
    }
  }

  override fun list(filter: FilenameFilter): Array<String> {
    return try {
      Files.list(_filePath).filter({ path: Path -> filter.accept(this, path.fileName.toString()) }).map({ obj: Path -> obj.toString() }).toArray{ arrayOf() }
    } catch(e: IOException) {
      arrayOf()
    }
  }

  override fun listFiles(): Array<File> {
    return try {
      Files.list(_filePath).map({ filePath: Path -> ProxyFile(filePath) }).toArray( { arrayOf() } )
    } catch(e: IOException) {
      arrayOf()
    }
  }

  override fun listFiles(filter: FileFilter): Array<File> {
    return try {
      Files.list(_filePath).map { filePath: Path -> ProxyFile(filePath) }.filter { pathname: ProxyFile? -> filter.accept(pathname) }.toArray( { arrayOf() } )
    } catch(e: IOException) {
      arrayOf()
    }
  }

  override fun listFiles(filter: FilenameFilter): Array<File> {
    return try {
      Files.list(_filePath).filter { path: Path -> filter.accept(this, path.fileName.toString()) }.map { filePath: Path -> ProxyFile(filePath) }.toArray( { arrayOf() } )
    } catch(e: IOException) {
      arrayOf()
    }
  }

  override fun mkdir(): Boolean {
    return try {
      Files.createDirectory(_filePath)
      true
    } catch(e: IOException) {
      false
    }
  }

  override fun mkdirs(): Boolean {
    return try {
      Files.createDirectories(_filePath) != null
    } catch(e: IOException) {
      false
    }
  }

  override fun renameTo(dest: File): Boolean {
    return try {
      Files.move(_filePath, dest.toPath())
      true
    } catch(e: IOException) {
      false
    }
  }

  override fun setExecutable(executable: Boolean): Boolean {
    return this.setExecutable(executable, true)
  }

  override fun setExecutable(executable: Boolean, ownerOnly: Boolean): Boolean {
    return try {
      val permissions: MutableSet<PosixFilePermission> = HashSet(Files.getAttribute(_filePath, "posix.permissions") as Set<PosixFilePermission>)
      if(executable) {
        val newPermissions = if(ownerOnly) EnumSet.of(PosixFilePermission.OWNER_EXECUTE) else EnumSet.of(PosixFilePermission.GROUP_EXECUTE, PosixFilePermission.OTHERS_EXECUTE, PosixFilePermission.OWNER_EXECUTE)
        permissions.addAll(newPermissions)
      } else {
        permissions.remove(PosixFilePermission.OWNER_EXECUTE)
        if(!ownerOnly) {
          permissions.remove(PosixFilePermission.GROUP_EXECUTE)
          permissions.remove(PosixFilePermission.OTHERS_EXECUTE)
        }
      }
      Files.setAttribute(_filePath, "posix.permissions", permissions)
      true
    } catch(e: IOException) {
      false
    }
  }

  override fun setLastModified(time: Long): Boolean {
    return try {
      Files.setLastModifiedTime(_filePath, FileTime.fromMillis(time))
      true
    } catch(e: IOException) {
      false
    }
  }

  override fun setReadable(readable: Boolean): Boolean {
    return this.setReadable(readable, true)
  }

  override fun setReadable(readable: Boolean, ownerOnly: Boolean): Boolean {
    return try {
      val permissions: MutableSet<PosixFilePermission> = HashSet(Files.getAttribute(_filePath, "posix.permissions") as Set<PosixFilePermission>)
      if(readable) {
        val newPermissions = if(ownerOnly) EnumSet.of(PosixFilePermission.OWNER_READ) else EnumSet.of(PosixFilePermission.GROUP_READ, PosixFilePermission.OTHERS_READ, PosixFilePermission.OWNER_READ)
        permissions.addAll(newPermissions)
      } else {
        permissions.remove(PosixFilePermission.OWNER_READ)
        if(!ownerOnly) {
          permissions.remove(PosixFilePermission.GROUP_READ)
          permissions.remove(PosixFilePermission.OTHERS_READ)
        }
      }
      Files.setAttribute(_filePath, "posix.permissions", permissions)
      true
    } catch(e: IOException) {
      try {
        Files.setAttribute(_filePath, "dos.readonly", readable)
        true
      } catch(e1: IOException) {
        false
      }
    }
  }

  override fun setReadOnly(): Boolean {
    return try {
      val newPermissions: Set<PosixFilePermission> = EnumSet.of(PosixFilePermission.GROUP_READ, PosixFilePermission.OTHERS_READ, PosixFilePermission.OWNER_READ)
      val permissions: MutableSet<PosixFilePermission> = HashSet(Files.getAttribute(_filePath, "posix.permissions") as Set<PosixFilePermission>)
      permissions.addAll(newPermissions)
      Files.setAttribute(_filePath, "posix.permissions", permissions)
      true
    } catch(e: IOException) {
      try {
        Files.setAttribute(_filePath, "dos.readonly", java.lang.Boolean.TRUE)
        true
      } catch(e1: IOException) {
        false
      }
    }
  }

  override fun setWritable(writable: Boolean): Boolean {
    return this.setWritable(writable, true)
  }

  override fun setWritable(writable: Boolean, ownerOnly: Boolean): Boolean {
    return try {
      val permissions: MutableSet<PosixFilePermission> = HashSet(Files.getAttribute(_filePath, "posix.permissions") as Set<PosixFilePermission>)
      if(writable) {
        val newPermissions = if(ownerOnly) EnumSet.of(PosixFilePermission.OWNER_WRITE) else EnumSet.of(PosixFilePermission.GROUP_WRITE, PosixFilePermission.OTHERS_WRITE, PosixFilePermission.OWNER_WRITE)
        permissions.addAll(newPermissions)
      } else {
        permissions.remove(PosixFilePermission.OWNER_WRITE)
        if(!ownerOnly) {
          permissions.remove(PosixFilePermission.GROUP_WRITE)
          permissions.remove(PosixFilePermission.OTHERS_WRITE)
        }
      }
      Files.setAttribute(_filePath, "posix.permissions", permissions)
      true
    } catch(e: IOException) {
      false
    }
  }

  override fun toPath(): Path {
    return _filePath
  }

  override fun toString(): String {
    return _filePath.toString()
  }

  override fun toURI(): URI {
    return _filePath.toUri()
  }

  companion object {
    private const val serialVersionUID = 1L
  }

}
